src/ray/object_manager/memory_object_reader.cc

```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/object_manager/memory_object_reader.h"
+#include <cstring>
+
+namespace ray {
+
+MemoryObjectReader::MemoryObjectReader(plasma::ObjectBuffer object_buffer,
+                                       rpc::Address owner_address)
+    : object_buffer_(std::move(object_buffer)),
+      owner_address_(std::move(owner_address)) {}
+
+uint64_t MemoryObjectReader::GetDataSize() const { return object_buffer_.data->Size(); }
+
+uint64_t MemoryObjectReader::GetMetadataSize() const {
+  return object_buffer_.metadata->Size();
+}
+
+const rpc::Address &MemoryObjectReader::GetOwnerAddress() const { return owner_address_; }
+
+bool MemoryObjectReader::ReadFromDataSection(uint64_t offset, uint64_t size,
+                                             char *output) const {
+  if (offset + size > GetDataSize()) {
+    return false;
+  }
+  std::memcpy(output, object_buffer_.data->Data() + offset, size);
```

Comment:
Isn't this regression? I remember we talked about that the plasma object is not copied until it is copied to the grpc request, but now it seems like we are copying twice? (plasma -> chunk -> grpc)?

Reply:
in the new code we only copy once too! because we moved to grpc instead of copy:
https://github.com/ray-project/ray/pull/16976/files#diff-1b792600903619ac4375f3eaf4cfe68f30cb8df813128d2a6fbee968a656554bR520

Reply:
added two tests to demonstrate the new code and the old code both only copy once.